ls1intum/Artemis
# [Bugfix]  Set submits of an exam programming exercise after the individual end date to ILLEGAL

### Checklist

### Motivation and Context


In case a student has still the write permission of his repository of the exam programming exercise, a student is able to submit code and receive a valid and rated result.

### Description

Directly after the commit we check if the exercise is an exam exercise and if the individual end date of that student is over. If it is over we set the submission to `ILLEGAL` and notify the instructor about the "cheating". The created result in the CIS Service is then set to `rated = false`.

These changes currently only affect the programming exercise participation for an exam.

### Note: Ignore for now: TODOs which will be done in a Follow-Up PR:
- filter for `ILLEGAL` submissions in all submissions and all participation results
  - this needs to be done in the DB
  - and also on server side 
- during assessment if submission is `ILLEGAL` get the file-content of the repository commit which does not belong to the `ILLEGAL` submission! 
- fetch the correct build log via commit id instead of latest

-> we only want to show the `ILLEGAL` submissions within the Submission View.
-> The problem here is that we usually take the last submission of a participation as valid. With these changes (which handles the edge-case) it is possible that `manual` results are not stored in the latest submission.


### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create an exam with a programming exercise
4. Set the individual time for some students to have enough time to test after the official end date has passed
5. Participate with a student in the exam and submit, then either
5.1 hand in early,
5.2 wait until the end and submit the exam, 
5.3 wait until the end and not submit the exam,
6. Now submit code afterwards via a local git client 
7. All further submissions should have the state `ILLEGAL` and have `rated=false`
8. Test all possible occasions where Participations with Results and where Submissions are fetched from the server and verify everything works as expected. 
9. Assessment of the programming exercise should display the last rated and valid submission (this means correct repository checkout, correct build logs, correct result string)
10. The Submissions View shows all submissions including the `ILLEGAL` ones

### Test Coverage




### Screenshots


#### Instructor Notification:
![image](https://user-images.githubusercontent.com/33484318/112221152-67eb0200-8c27-11eb-91e5-7f20675d28f7.png)



#### Table shows only valid and graded results: 
![image](https://user-images.githubusercontent.com/33484318/111880057-71ffcd00-89a9-11eb-8c6d-f7ceb8c923a3.png)
#### The Submissions View also shows the INVALID submissions:
![image](https://user-images.githubusercontent.com/33484318/111880071-880d8d80-89a9-11eb-8224-60fd5b3535b4.png)

#### Student Results should be only the last valid results:
![image](https://user-images.githubusercontent.com/33484318/111880181-2a2d7580-89aa-11eb-8c3d-b369d861fd7b.png)

#### Example of the correct result string but the wrong repository commit checkout (The valid submission has 0 points and does not sort the input):
![image](https://user-images.githubusercontent.com/33484318/111910402-82737e80-8a61-11eb-84ef-2f14767697c9.png)
